test/extensions/transport_sockets/proxy_protocol/proxy_protocol_integration_test.cc

```diff
@@ -0,0 +1,185 @@
+          proxy_proto_transport;
+      proxy_proto_transport.mutable_transport_socket()->MergeFrom(inner_socket);
+      proxy_proto_transport.mutable_config()->MergeFrom(proxy_proto_config);
+      transport_socket->mutable_typed_config()->PackFrom(proxy_proto_transport);
+
+      if (health_checks_) {
+        auto health_check =
+            bootstrap.mutable_static_resources()->mutable_clusters(0)->add_health_checks();
+        health_check->mutable_interval()->set_seconds(15);
+        health_check->mutable_timeout()->set_nanos(100000000); // 100 ms
+        health_check->mutable_no_traffic_interval()->set_seconds(15);
+        health_check->mutable_unhealthy_threshold()->set_value(3);
+        health_check->mutable_healthy_threshold()->set_value(3);
+        health_check->mutable_tcp_health_check()->mutable_send()->set_text("434c4f53450a");
+        auto recv_payload = health_check->mutable_tcp_health_check()->mutable_receive()->Add();
+        recv_payload->set_text("4f4b");
+      }
+    });
+    BaseIntegrationTest::initialize();
+  }
+
+  FakeRawConnectionPtr fake_upstream_connection_;
+
+private:
+  envoy::config::core::v3::ProxyProtocolConfig_Version version_;
+  bool health_checks_;
+  std::string inner_socket_;
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, ProxyProtocolIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+// Test sending proxy protocol v1
+TEST_P(ProxyProtocolIntegrationTest, TestV1ProxyProtocol) {
+  setup(envoy::config::core::v3::ProxyProtocolConfig::V1, false,
+        "envoy.transport_sockets.raw_buffer");
+  initialize();
+
+  auto listener_port = lookupPort("listener_0");
+  auto tcp_client = makeTcpConnection(listener_port);
+  ASSERT_TRUE(fake_upstreams_[0]->waitForRawConnection(fake_upstream_connection_));
+
+  std::string observed_data;
+  ASSERT_TRUE(tcp_client->write("data"));
+  if (GetParam() == Network::Address::IpVersion::v4) {
+    ASSERT_TRUE(fake_upstream_connection_->waitForData(48, &observed_data));
+    EXPECT_THAT(observed_data, testing::StartsWith("PROXY TCP4 127.0.0.1 127.0.0.1 "));
+  } else if (GetParam() == Network::Address::IpVersion::v6) {
+    ASSERT_TRUE(fake_upstream_connection_->waitForData(36, &observed_data));
+    EXPECT_THAT(observed_data, testing::StartsWith("PROXY TCP6 ::1 ::1 "));
+  }
+  EXPECT_THAT(observed_data, testing::EndsWith(absl::StrCat(" ", listener_port, "\r\ndata")));
+
+  auto previous_data = observed_data;
+  observed_data.clear();
+  ASSERT_TRUE(tcp_client->write(" more data"));
+  ASSERT_TRUE(fake_upstream_connection_->waitForData(previous_data.length() + 10, &observed_data));
+  EXPECT_EQ(previous_data + " more data", observed_data);
+
+  tcp_client->close();
+  ASSERT_TRUE(fake_upstream_connection_->waitForDisconnect());
+}
+
+// Test header is sent unencrypted using a TLS inner socket
+TEST_P(ProxyProtocolIntegrationTest, TestTLSSocket) {
```

Comment:
This test was taking 5 seconds to run locally. I'm not sure how we can speed it up.

Reply:
Try an -l trace and see where it hangs?  Maybe it's hitting a timeout somewhere?  If this is all raw TCP I think half close semantics apply, so if you close the client the connection may persist until the idle timeout kicks in.  

if 
  ASSERT_TRUE(fake_upstream_connection->close());
  ASSERT_TRUE(fake_upstream_connection->waitForDisconnect());
  tcp_client->waitForHalfClose();
  tcp_client->close();
fixes, that's the problem, though I'm surprised there wouldn't be the same stall in other tests

Reply:
Thanks for the tips! Closing the upstream connection fixed it.